relay-general/src/pii/attachments.rs

```diff
@@ -110,6 +114,259 @@ fn apply_regex_to_bytes(
+            let char_len = if 0xD800 & code == 0xD800 {
+                size * 2 // leading surrogate
+            } else {
+                size
+            };
+            if (len - offset) < char_len {
+                break; // Not enough space for this char
+            }
+            let target = &mut self.as_bytes_mut()[offset..offset + size];
+            target.copy_from_slice(&code.to_le_bytes());
+            offset += size;
+        }
+
+        // Our current WStr could be invalid UTF-16LE at this point, up to the current
+        // offset it is valid, but afterwards is may not be.  So we can not just slice
+        // ourselves and must slice the raw bytes, creating a new WStr with them.  This new
+        // WStr could be invalid, but after we called .fill_content() it will be valid.
+        let remainder_bytes = &mut self.as_bytes_mut()[offset..];
+        let remainder = unsafe { WStr::from_utf16le_unchecked_mut(remainder_bytes) };
+        remainder.fill_content(padding)
+    }
+}
+
+struct MutSegmentIter<'a> {
+    data: &'a mut [u8],
+    decoder: Box<dyn RawDecoder>,
+    offset: usize,
+}
+
+impl<'a> MutSegmentIter<'a> {
+    fn new(data: &'a mut [u8], encoding: impl Encoding) -> Self {
+        Self {
+            data,
+            decoder: encoding.raw_decoder(),
+            offset: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for MutSegmentIter<'a> {
+    type Item = MutSegment<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut decoded = String::with_capacity(self.data.len() - self.offset);
+
+        loop {
+            if self.offset >= self.data.len() {
+                return None;
+            }
+
+            decoded.clear();
+            let start = self.offset;
+            let (unprocessed_offset, err) =
+                self.decoder.raw_feed(&self.data[start..], &mut decoded);
+            let end = start + unprocessed_offset;
+
+            if let Some(err) = err {
+                if err.upto > 0 {
+                    self.offset += err.upto as usize;
+                } else {
+                    // This should never happen, but if it does, re-set the decoder and skip
+                    // forward to the next 2 bytes.
+                    self.offset += std::mem::size_of::<u16>(); // TODO: encoding-neutral?!?
+                    self.decoder = self.decoder.from_self();
+                }
+                if decoded.len() > 2 {
+                    return Some(MutSegment {
+                        raw: unsafe { std::mem::transmute(&mut self.data[start..end]) },
+                        decoded,
+                    });
+                } else {
+                    continue;
+                }
+            } else {
+                self.offset += unprocessed_offset;
+                if decoded.len() > 2 {
+                    return Some(MutSegment {
+                        raw: unsafe { std::mem::transmute(&mut self.data[start..end]) },
```

Comment:
Are we transmuting only the lifetimes here? In any case i'd explicitly specify the type parameters of transmute to make this more readable. Otherwise random stuff is inferred if the struct def is changed.

Reply:
Also I think i'd prefer making this a function that takes a callback. instead of this `loop { }` and unsafe usage.

Reply:
i made the transmute explicit.

I prefer the iterator over a callback i think.  But the loop would sadly stay though, it's a result of skipping over strings that are only 1 or 2 characters long.  I think it could be a reasonable call to not skip those, our regexes would either match or not anyway and then the loop can go.

Reply:
So this loop is still an interesting design decision.  I think we should remove it and try to match regexes even if we only have 1 or 2 characters, in normal cases the regex would just not match.  In the cases where it would match it is a rather invisible bit of weird behaviour that we'd skip over it.

Plus this code would be simpler and safer :)

Reply:
keep it, and make it 5